File: medium/3_longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        curr_list = []
        max_len = 0
        for char in s:
            if char in curr_list:
                if len(curr_list) > max_len:
                    max_len = len(curr_list)
                curr_list = curr_list[curr_list.index(char)+1:]
            curr_list.append(char)
        if len(curr_list) > max_len:
            return len(curr_list)
        return max_len

File: medium/3_longest_substring_without_repeating_characters.go
func lengthOfLongestSubstring(s string) int {
	curr_set := map[rune]struct{}{}
	var curr_slice []rune
	max_len := 0
	for _, char := range s {
		if _, ok := curr_set[char]; ok {
			if len(curr_slice) > max_len {
				max_len = len(curr_slice)
			}
			for _, c := range curr_slice {
				curr_slice = curr_slice[1:]
				delete(curr_set, c)
				if c == char {
					break
				}
			}
		}
		curr_slice = append(curr_slice, char)
		curr_set[char] = struct{}{}
	}
	if len(curr_slice) > max_len {
		return len(curr_slice)
	}
	return max_len
}
